File: RecursiveFilter.cs
using System;
using System.Collections.Generic;
using System.Linq;

class RecursiveFilter
{
    // 递归过滤，去掉超出均值 ±k*std 的数据，直到所有数据都落在范围内。
    public static FilterResult RecursiveFilterFunc(List<double> data, double k)
    {
        var removed = new List<double>();
        var current = new List<double>(data);
        while (true)
        {
            double mean = current.Average();
            double std = Std(current, mean);
            var filtered = current.Where(x => x >= mean - k * std && x <= mean + k * std).ToList();
            if (filtered.Count == current.Count)
                return new FilterResult(filtered, mean, std, removed);
            foreach (var x in current)
            {
                if (x < mean - k * std || x > mean + k * std)
                    removed.Add(x);
            }
            current = filtered;
        }
    }

    public static double Std(List<double> data, double mean)
    {
        double variance = data.Select(x => (x - mean) * (x - mean)).Average();
        return Math.Sqrt(variance);
    }

    public static void PrintFilterResult(FilterResult result, double k)
    {
        Console.WriteLine("最终保留数据: " + string.Join(", ", result.Filtered));
        Console.WriteLine("最终平均值 = {0:F2}", result.Mean);
        Console.WriteLine("最终标准差 = {0:F2}", result.Std);
        Console.WriteLine("{0}倍标准差上下线: [{1:F2}, {2:F2}]", k, result.Mean - k * result.Std, result.Mean + k * result.Std);
        Console.WriteLine("被过滤掉的异常值: " + string.Join(", ", result.Removed));
    }

    static void Main(string[] args)
    {
        var volumes = new List<double> { 190, 191, 181, 119, 166, 90, 123, 131, 356, 1057 };
        double k = 2.0;
        var result = RecursiveFilterFunc(volumes, k);
        PrintFilterResult(result, k);
    }

    public class FilterResult
    {
        public List<double> Filtered;
        public double Mean;
        public double Std;
        public List<double> Removed;
        public FilterResult(List<double> filtered, double mean, double std, List<double> removed)
        {
            Filtered = filtered;
            Mean = mean;
            Std = std;
            Removed = removed;
        }
    }
}

File: RecursiveFilter.java
import java.util.*;
import java.util.stream.Collectors;

public class RecursiveFilter {
    /**
     * 递归过滤，去掉超出均值 ±k*std 的数据，直到所有数据都落在范围内。
     * 返回过滤后数据、均值、标准差、被过滤掉的异常值
     */
    public static FilterResult recursiveFilter(List<Double> data, double k) {
        List<Double> removed = new ArrayList<>();
        List<Double> current = new ArrayList<>(data);
        while (true) {
            double mean = mean(current);
            double std = std(current, mean);
            List<Double> filtered = current.stream()
                    .filter(x -> x >= mean - k * std && x <= mean + k * std)
                    .collect(Collectors.toList());
            if (filtered.size() == current.size()) {
                return new FilterResult(filtered, mean, std, removed);
            }
            for (double x : current) {
                if (x < mean - k * std || x > mean + k * std) {
                    removed.add(x);
                }
            }
            current = filtered;
        }
    }

    public static double mean(List<Double> data) {
        return data.stream().mapToDouble(Double::doubleValue).average().orElse(0.0);
    }

    public static double std(List<Double> data, double mean) {
        double variance = data.stream().mapToDouble(x -> (x - mean) * (x - mean)).average().orElse(0.0);
        return Math.sqrt(variance);
    }

    public static void printFilterResult(FilterResult result, double k) {
        System.out.println("最终保留数据: " + result.filtered);
        System.out.printf("最终平均值 = %.2f\n", result.mean);
        System.out.printf("最终标准差 = %.2f\n", result.std);
        System.out.printf("%.0f倍标准差上下线: [%.2f, %.2f]\n", k, result.mean - k * result.std, result.mean + k * result.std);
        System.out.println("被过滤掉的异常值: " + result.removed);
    }

    public static void main(String[] args) {
        List<Double> volumes = Arrays.asList(190.0, 191.0, 181.0, 119.0, 166.0, 90.0, 123.0, 131.0, 356.0, 1057.0);
        double k = 2.0;
        FilterResult result = recursiveFilter(volumes, k);
        printFilterResult(result, k);
        // 如需绘图，可导出数据用Python等工具绘制
    }

    static class FilterResult {
        List<Double> filtered;
        double mean;
        double std;
        List<Double> removed;
        FilterResult(List<Double> filtered, double mean, double std, List<Double> removed) {
            this.filtered = filtered;
            this.mean = mean;
            this.std = std;
            this.removed = removed;
        }
    }
}
